File: app/src/main/java/io/github/superbderrick/scoreboard/helper/ScoreManager.java
package io.github.superbderrick.scoreboard.helper;

import android.util.Log;

/**
 * Created by derrick on 27/01/2018.
 */

public class ScoreManager {

    public interface OnScoreChangeListener {
        public void onFirstScoreChanged(int score);
        public void onSecondScoreChanged(int score);
    }

    public enum Operation { Increase, Decrease}
    public enum UserType { First, Second}

    public static final int DEFAULT_MAXIMUM_SCORE = 30;

    private int mFirstScore , mSecondScore;
    private int mScoreMaxRange;
    private OnScoreChangeListener mListener;

    public OnScoreChangeListener getListener() {
        return mListener;
    }

    public void setListener(OnScoreChangeListener mScoreListener) {
        this.mListener = mScoreListener;
    }

    public int getScoreMaxRange() {
        return mScoreMaxRange;
    }
    public void setScoreMaxRange(int mScoreMaxRange) {
        this.mScoreMaxRange = mScoreMaxRange;
    }

    public int getFirstScore() {
        return mFirstScore;
    }

    public int getSecondScore() {
        return mSecondScore;
    }

    public void changeScore(Operation operation , UserType type) {

        if(type == UserType.First) {
            regulateFirstScore(operation);
        } else {
            regulateSecondScore(operation);
        }
    }

    private void regulateFirstScore(Operation operation) {

        if(operation == Operation.Increase) {
            if (mFirstScore < mScoreMaxRange && mFirstScore >=0) {
                mFirstScore ++;
            }
        } else {
            if(mFirstScore > 0 && mFirstScore < mScoreMaxRange+1) {
                mFirstScore --;
            }
        }

        mListener.onFirstScoreChanged(mFirstScore);

    }

    private void regulateSecondScore(Operation operation) {
        if(operation == Operation.Increase) {
            if (mSecondScore < mScoreMaxRange && mSecondScore >=0) {
                mSecondScore ++;
            }
        } else {
            if(mSecondScore > 0 && mSecondScore < mScoreMaxRange+1) {
                mSecondScore --;
            }
        }

        mListener.onSecondScoreChanged(mSecondScore);

    }

    public void resetScore() {
        mFirstScore = 0;
        mSecondScore = 0;
    }


    public ScoreManager() {
        initValues();
    }

    private void initValues() {
        mFirstScore = 0;
        mSecondScore = 0;
    }
}

File: app/src/main/java/io/github/superbderrick/scoreboard/helper/ScoreManager.kt
package io.github.superbderrick.scoreboard.helper

/**
 * Created by derrick on 27/01/2018.
 */
class ScoreManager {
    interface OnScoreChangeListener {
        fun onFirstScoreChanged(score: Int)
        fun onSecondScoreChanged(score: Int)
        fun onScoreInitialized(score: Int)
    }

    enum class Operation {
        Increase, Decrease
    }

    enum class UserType {
        First, Second
    }

    var firstScore = 0
        private set
    var secondScore = 0
        private set
    var scoreMaxRange = 0
    var listener: OnScoreChangeListener? = null

    fun changeScore(operation: Operation, type: UserType) {
        if (type == UserType.First) {
            regulateFirstScore(operation)
        } else {
            regulateSecondScore(operation)
        }
    }

    fun changeScore(operation: Operation, type: UserType, addtionalPoint: Int) {
        if (type == UserType.First) {
            regulateFirstScore(operation, addtionalPoint)
        } else {
            regulateSecondScore(operation, addtionalPoint)
        }
    }

    private fun regulateFirstScore(operation: Operation) {
        if (operation == Operation.Increase) {
            if (firstScore < scoreMaxRange && firstScore >= 0) {
                firstScore++
            }
        } else {
            if (firstScore > 0 && firstScore < scoreMaxRange + 1) {
                firstScore--
            }
        }
        listener!!.onFirstScoreChanged(firstScore)
    }

    private fun regulateFirstScore(operation: Operation, addtionalPoint: Int) {
        if (operation == Operation.Increase) {
            if (firstScore < scoreMaxRange && firstScore >= 0) {
                firstScore = firstScore + addtionalPoint
            }
        } else {
            if (firstScore > 0 && firstScore < scoreMaxRange + 1) {
                firstScore--
            }
        }
        listener!!.onFirstScoreChanged(firstScore)
    }

    private fun regulateSecondScore(operation: Operation) {
        if (operation == Operation.Increase) {
            if (secondScore < scoreMaxRange && secondScore >= 0) {
                secondScore++
            }
        } else {
            if (secondScore > 0 && secondScore < scoreMaxRange + 1) {
                secondScore--
            }
        }
        listener!!.onSecondScoreChanged(secondScore)
    }

    private fun regulateSecondScore(operation: Operation, addtionalPoint: Int) {
        if (operation == Operation.Increase) {
            if (secondScore < scoreMaxRange && secondScore >= 0) {
                secondScore = secondScore + addtionalPoint
            }
        } else {
            if (secondScore > 0 && secondScore < scoreMaxRange + 1) {
                secondScore--
            }
        }
        listener!!.onSecondScoreChanged(secondScore)
    }

    fun resetScore() {
        firstScore = 0
        secondScore = 0
        listener!!.onScoreInitialized(0)
    }

    private fun initValues() {
        firstScore = 0
        secondScore = 0
    }



    companion object {
        const val DEFAULT_MAXIMUM_SCORE = 30
    }

    init {
        initValues()
    }
}